### DISCUSSION TITLE
A different approach to Hello World!

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#987] [Getting started]
As I'm struggling through getting started with moddable, I'm pondering how it compares with other experiences I've had... What if "getting started" was turned around a bit? I'll take the esp32 as example, but I'm sure the same could be done with Pico.

1. Download an esp32 image that has: mod base, wifi stuff, http stuff, REPL
2. Install esptool (not esp-idf) and miniterm.py or equiv
3. Flash esp32
4. Connect to it using miniterm and be greeted by the REPL
5. Define a function and run it using a timer to blink an LED (have to start with that I guess)
6. Type a couple of commands to set-up Wifi
7. Type a function to download some weather info (or something else) and print the result

Some additional little examples could, I'm sure, be adapted from the existing ones.

The next step would be loading some mods. I'm assuming that only the XS tools are needed for that, none of esp-idf. I don't know how much of the Moddable SDK is required, something easy to install would be key. Or maybe that's an MSI/AppImage/...

Either way:
1. Copy a mod that serves up a little web page with a toggle to turn the LED on/off
2. Run a command to compile
3. Run a command to install & run on the esp32
4. Play
5. Alternatively do this coupled with xsbug

I believe a getting started path like the above has the following properties:
- shortest path to hello world on the device
- makes it easy to do some simple experimentation without needing a toolchain and all that
- stays firmly in the mostly interactive javascript world
- could allow significant projects without having to get the whole Moddable SDK nor ESP-IDF

I realize that to those of you that breathe the SDK day in and day out this may not seem to be that different from today's getting started and the large tree of examples, but I believe it does present a very different picture.

**Comment by phoddie** [comment]
There isn't a single "getting started" experience that is the right fit for everyone. There are different onramps depending on the goal. You are describing one. By coincidence, our [book](https://www.moddable.com/book) works pretty much like what you describe (except you build the image instead of downloading it). 

What you are describing is a "mods-only" experience. That is great for some folks. If you'd like to write that up, that would be great.

**Reply by tve** [reply]
Part of the inspiration came from seeing [this thread](https://forums.raspberrypi.com/viewtopic.php?t=304964) on the rPi forum where people are clearly looking for something simpler to get started with and with an interactive REPL...

Couple of motivating examples from personal experience:
- I just tried the lora example and it errors on me #989 with `get device: undefined variable`. If I was using Espruino or Micropython I'd just type `device` into the REPL to see whether that's supposed to be a global or what and go from there.
- I've built many Micropython images, I have a pile of build envs on my system, when v1.19 was released I wanted to quickly try out some new features, rather than going through the hour downloading and installing mpy 1.19 and esp-idf and all that I just loaded the standard image on an esp32 and typed some commands, plus uploaded a couple of files to try stuff.

**Comment by phoddie** [comment]
Thank for your sharing your inspiration.

As I noted, much of what you describe has been possible for some time. If you'd like to help pull together a system to maintain updated images, build mod versions of examples, that would be fantastic. From what you describe, this seems well within your abilities. And you are clearly passionate about it. I'm enthusiastic to support that.

Please do keep in mind that there are many different ways developers use the Moddable SDK. This isn't a question of a right or wrong approach. The approach you outline would be great for one group of developers, but we do have groups working in other, very different ways.

**Reply by tve** [reply]
I totally understand that there are other approaches and what attracts me to XS in the first place is some of those other things, like compiling and optimizing on the host and a less REPL centric approach. I'm pretty stumped by having now spent three days without being able to show anything other than the blink example for it... Maybe if I was just using a mac it would all be smooth cruising... Or maybe I need to order the book, wait for Amazon to deliver ("dec 22nd to 29th"), and then remember what Moddable is... ;-) )

**Reply by phoddie** [reply]
> I'm pretty stumped...

You seem to be exploring far and wide. Perhaps a more focused approach to getting started would be more effective.

> Maybe if I was just using a Mac...

We have developers working on Mac, Windows, and Linux. Linux is generally the toughest because there are so many different configurations -- no project manages to support them all.

> Or maybe I need to order the book, wait for Amazon to deliver ("dec 22nd to 29th"), and then remember what Moddable is... ;-) )

FWIW – our publisher offers an[ eBook](https://link.springer.com/book/10.1007/978-1-4842-5070-9) version. Apparently that is also available for Kindle, if you prefer. Either way, that should arrive in seconds, ;)
